### DISCUSSION TITLE
FPU Configuration

### DISCUSSION THREAD

**Discussion by rodrigobrochado** [zephyrproject-rtos/zephyr#29174] [Q&A]
I have 2 cooperative threads that use float variables to do some computations. I'm a little confused if I need to set up the [Floating Point Services](https://docs.zephyrproject.org/2.4.0/reference/kernel/other/float.html). I'm not sure if it is only required for low-level operations with FP registers or for normal math operations as well.  

Because I didn't set it up (and have no CONFIG_FPU on autoconf.h), both of my threads run ok and I don't get any errors, although the doc says that "the kernel generates a fatal error condition and aborts the thread" if I use FP registers in [No FP registers mode](https://docs.zephyrproject.org/2.4.0/reference/kernel/other/float.html#no-fp-registers-mode).

The fatal error would be only for preemptive threads?

Thanks,

**Comment by henrikbrixandersen** [comment]
Are you using an Arm Cortex-m with FPU? If so, any threads is allowed to access the FP registers: https://docs.zephyrproject.org/2.4.0/reference/kernel/other/float.html#arm-cortex-m-architecture-with-the-floating-point-extension

If not, please share which SoC you are using. Thanks. 

The documentation, you linked to, is for Zephyr v2.4.0. Is this the version you are using?

**Reply by rodrigobrochado** [reply]
Thanks for your reply @henrikbrixandersen .

Yes, I'm using 2.4.0 and SoC nrf52840 (Cortex-M). 

Ok, it's allowed to access the FP registers, but is it automatically activated?

Also, access to the registers happens behind the scenes, right? If I made some float computations (I don't use the registers manually) without setting the FP flags on the thread, shouldn't it throw a fatal error?

**Reply by henrikbrixandersen** [reply]
According to the documentation, I linked to, it is automatically activated:

> On the ARM Cortex-M architecture with the Floating Point Extension, the kernel treats all threads as FPU users when shared FP registers mode is enabled. This means that any thread is allowed to access the floating point registers. The ARM kernel automatically detects that a given thread is using the floating point registers the first time the thread accesses them.

**Reply by rodrigobrochado** [reply]
Humm, I think this highlighted part is open to other interpretations:  
> the kernel treats all threads as FPU users **when shared FP registers mode is enabled**

I understood that it required the flag to be previously set, like using K_FP_REGS on `k_thread_create()` (that is mentioned in the next paragraph of the part quoted).  

So when using Cortex-M devices, I don't need to set anything, like CONFIG_FPU or K_FP_REGS, etc?
